modules: The 'features' module

seiko2plus opened the pull request:
This module aims to provide a more dynamic way to control common
  features between compilers that would generally require special
  headers, arguments, or compile-time test cases, quite ideal
  to manage CPU features and solve compatibility issues.

TODO:
- ~[ ] add x86 features~
- ~[ ] add Arm features~
- ~[ ] add PPC64 features~
- ~[ ] add IBMZ features~
- [x] add static testing unit
- [ ] add document
- [x] improve log/debug
- [x] improve the performance, e.g. LRU

seiko2plus committed:
modules: The 'feature' module

  This module aims to provide a more dynamic way to control common
  features between compilers that would generally require special
  headers, arguments, or compile-time test cases, quite ideal
  to manage CPU features and solve compatibility issues.

seiko2plus committed:
modules: Add method 'feature.cpu_features'

  and brings X86 CPU features

seiko2plus committed:
Fix runtime detect of AVX features

seiko2plus committed:
Satisfy the linter

seiko2plus committed:
replace method to_dict with get

seiko2plus committed:
improve test method

seiko2plus committed:
Fix returned boolean value when test fails

rgommers commented:
Looks promising, thanks @seiko2plus!

One question I have is what CPU features are of interest here aside from SIMD ones? From the discussion at https://github.com/mesonbuild/meson/discussions/11033 I thought you were going to try and fit this into the existing `simd` module. I don't personally have a preference here - whatever works and has a good API I'd say.

For others, @seiko2plus posted this test program to try this out earlier ([here](https://github.com/numpy/numpy/issues/22546#issuecomment-1386341562)):

```meson
project('test', 'c')

feature = import('feature')
cpu_features = feature.cpu_features()

targets = [['AVX2_FMA3', [cpu_features['AVX2'], cpu_features['FMA3']]],
          ['AVX512_COMMON', cpu_features['AVX512_COMMON']],
          ['AVX512_SKX', cpu_features['AVX512_SKX']]]

foreach target : targets
  t = feature.test(target[1])
  if t[0]
    message('target name:', target[0])
    data = t[1]
    message('Implicit features:', data['implicit'])
    message('Compiler arguments:', data['args'])
    message('Headers:', data['headers'])
    message('Need to detect on runtime:', data['detect'])
    message('Supported #definitions:', data['defines'])
    message('Unsupported #definitions:', data['undefines'])
    message('Compiler support this target by ', data['support'])
    message('========================================')
  endif
endforeach
```

@seiko2plus this diff fixes the minor issue I had with feature detection here:

<details>

```diff
--- a/mesonbuild/modules/feature/x86_features.py
+++ b/mesonbuild/modules/feature/x86_features.py
@@ -42,7 +42,7 @@ if T.TYPE_CHECKING:
 
 def test_code_main(code: str) -> str:
     return dedent(f'''\
-        int main(int, char **argv)
+        int main(int unused, char **argv)
         {{
            char *src = argv[1];
            {dedent(code)}
```

</details>

seiko2plus commented:
> #11033 I thought you were going to try and fit this into the existing simd module.

The current meson SIMD module lacks flexibility and modifies it to support our needs in a way to provide static and dynamic dispatching (baseline/dispatch) that fits "any project needs" will require more effort than I expected.

So I just split the work into two modules rather than a single module covering all the operations, the first module is the module 'feature' which provides what is necessary to be coded by python, and the second module 'compiler_opt' will be left for future step basically it will require to convert the following two meson scripts (not completed yet) into python module:
  - CPU dispatcher initializer
  - Dispatch-able sources builder

We still can modify the current meson SIMD module to allow only the use of the CPU features in this module without any extra modifications but we're not going to use it in NumPy, for many considerations are hard to explain in one comment.

> One question I have is what CPU features are of interest here aside from SIMD ones?

Do you mean aside from SIMD features? Any CPU features in general, but main concern features that will be required for optimizing certain operations. For example, POPCNT and LZCNT.  

But if you mean aside from the current meson SIMD module? 

Flexibility is the main reason, imagine if a project wants to modify or add a new feature or a new architecture. What if there is a necessity to make fast or custom modifications to an existing feature? some examples:

#### So assuming adding feature X implies feature `AVX512_ICL`:
```Meson
module_feature = import('feature')
cpu_features = module_feature.cpu_features()
icl = cpu_features['AVX512_ICL']
nfet = module_feature.new('X',  icl.get('interest') + 1, implies: icl, test_code='''
// testing code
''')
if cc.get_id() == 'gcc'
   nfet.update(args: '-mX') # flags
else
   nfet.update(disable: 'requires gcc')
endif
cpu_features += {'X': nfet}
``` 

#### What about modifying or adding a certain flag for a certain feature?

We faced an incident here https://github.com/numpy/numpy/pull/20991 which requires disabling `MMX` registers as a workaround.
```Meson
if cc.get_id() == 'gcc'
  fet = cpu_features['AVX512_COMMON']
  fet.update(args: fet.get('args') + '-mno-mmx')
endif
````
Now all the features that imply `AVX512_COMMON` have the new arg including feature `AVX512_COMMON`. I will try to cover most of the usage in the document that will come with this module.

rgommers commented:
> and the second module 'compiler_opt' will be left for future step basically it will require to convert the following two meson scripts (not completed yet) into python module:

I think that'd be a vendorable package containing only `meson.build` files (and maybe some C/C++ code), which one could put in a `subprojects` directory, rather than a Python package, right? There is nothing Python-specific about this I think. I'm not sure if there is precedent for that kind of thing in Meson-using projects, but I think it seems like a good plan. At least for now - it's pretty complex, so generalizing it from use by only NumPy to multiple other projects first makes sense to me.

> Do you mean aside from SIMD features? Any CPU features in general, but main concern features that will be required for optimizing certain operations. For example, POPCNT and LZCNT.

I meant SIMD features in general. Which is still the main use case here I think. But yes, `popcnt` and `lzcnt` is a good example, thanks.

> I will try to cover most of the usage in the document that will come with this module.

Nice examples you have given here already, thanks.

eli-schwartz commented:
> I think that'd be a vendorable package containing only `meson.build` files (and maybe some C/C++ code), which one could put in a `subprojects` directory, rather than a Python package, right? There is nothing Python-specific about this I think. I'm not sure if there is precedent for that kind of thing in Meson-using projects, but I think it seems like a good plan. At least for now - it's pretty complex, so generalizing it from use by only NumPy to multiple other projects first makes sense to me.

It certainly would not be the only meson-using project that used a subproject containing sources which are intended strictly for use as a vendorable package, together with the `meson.build` instructions for building it into a static library and yielding that to the superproject.

9243659864598 commented:
qggq

seiko2plus committed:
Shrink the module

  - Removes CPU features implmentation and move them NumPy meson scripts

  - Removes attr 'header', since it can be handeled within
    a configuration file
  - re-implment method test, make it more simple

  - Implment method `multi_targets`
    Handels multi_targets via meson script was pretty anoyning
    from python make it much simpler

seiko2plus committed:
Fix config src dir

seiko2plus committed:
fix initialize Compiler when user_defined_options is None

seiko2plus committed:
Several improvements

  - Improve multi_targets() method:

    * rename multi_target() to multi_targets() use plural to indicates handling more than one CPU target
    * Split the function into multiple smaller functions for better organization.
    * Modify the function to return a single library array instead of a list of libraries.
    * Avoid returning the result of testing the enabled features to keep it
    * simple.
    * Enhance debugging capabilities by printing the test results of enabled targets.
    * Avoid returns static_lib instead wraps it with another object
      to reduce the number of final generated objects also
      to sort the all objects based on lowest interest
    * validate the sort of dispatch features

  - fix mypy
  - init the module doc
  - allow disables the cache
  - fix method `features.new` name during trigger arguments errors
  - generate config files inside the builddir instead
  - add test cases
  - cache tests within coredata
  - fix duplicate baseline build
  - refactor module name from `feature` to `features`
  - fix python typying

seiko2plus committed:
expose CompileResult because its needed by unittest

seiko2plus committed:
Avoid full importing of module typing

seiko2plus commented:
@eli-schwartz, Hi Eli, I am curious if there a method to directly generate documentation for the supported methods from the Python scripts themselves.

eli-schwartz commented:
> @eli-schwartz, Hi Eli, I am curious if there a method to directly generate documentation for the supported methods from the Python scripts themselves.

Historically, no. All the documentation is either handwritten or produced as structured yaml in docs/yaml/. We do have a way to generate module documentation from structured yaml, but it's currently disabled because those were never properly ported (doing this would be nice, but also probably too complicated to hold up the 'feature' module over).

seiko2plus committed:
DOC: Improve the overview, functions signatures

seiko2plus commented:
> Historically, no. All the documentation is either handwritten or produced as structured yaml in docs/yaml/. We do have a way to generate module documentation from structured yaml, but it's currently disabled because those were never properly ported (doing this would be nice, but also probably too complicated to hold up the 'feature' module over).

Thank you for the clarification, its seem I will go for handwritten but I just realized different formats when it comes to methods signatures and module objects. Could you take a look at manual format that this module use, and suggest format for documenting module objects?

seiko2plus committed:
BUG: Fix features tests on cross-compile builds

eli-schwartz commented:
You could use the python module as partial inspiration: https://mesonbuild.com/Python-module.html#python_installation-object

It is a subheader with a list of methods in further subheaders, which aren't substantially different from module functions.

eli-schwartz commented:
A bit of vague high-level commentary:
- there is already a feature (options) object, adding another FeatureObject runs the risk of the two being confused. It's also not entirely clear what an `import('features')` does. Features of what? I would probably rename the module "cpufeatures".
- In the ModuleInfo, you can pass `unstable=True`, which will mark the module as provisional. This means that meson reserves the right to change how it works (or in theory even remove it) without falling under the usual semver stability policy. I think it would be a good idea to start off using this, with the understanding that numpy is the primary stakeholder in the module, as it gives us the leeway to rethink the design. Hopefully this won't be necessary -- but it is a pretty complex new module, so you never know. The only practical effect this will have on consumers such as numpy is a) logging a nonfatal warning "btw this is unstable, compatibility not guaranteed", b) if such changes ever do occur, one would need to pay close attention to which versions of meson are allowed by `meson.build`. Considering you'd effectively be the codeowners in meson for it, this isn't really a burden at all.

eli-schwartz commented:
I think this module would also probably merit a tutorial walkthrough, e.g. `docs/markdown/Using-CPU-multifunction-dispatch.md`.

For bonus points it would compare and contrast both this module and the existing (also unstable) SIMD module. This module is a lot more powerful, so it would be interesting to see how rewriting a SIMD module example would look here -- and by the same token, offer guidance on migrating over to the new module so that we could perhaps deprecate the SIMD module.

Vogtinator committed:
feature module: Fix handling of multiple conflicts per attribute

- Attributes without match were never actually added to the list
- Only the last conflict actually had an effect, earlier results were
  discarded

mattip commented:
Also see mesonbuild/meson#11885 which also proposes a module/features module

dcbaker commented:
I don't know that it's import, I can always rename the module in 11885 if we ever have agreement on adding that module, it could be named `options` for example; and this PR is much closer to landing than that one. So, as the author of 11885, I don't think it's worth worrying about

bruchar1 commented on test cases/features/1 baseline/init_features:
```diff
@@ -0,0 +1 @@
+../init_features
```
symlinks will be problematic on Windows platform.

tristan957 commented:
Seems like this is pretty close to getting merged. Would be great to get in for 1.6.

dcbaker commented on docs/markdown/Features-module.md:
```diff
@@ -0,0 +1,320 @@
+
+## Overview
+
+Dealing with a numerous of CPU features within C and C++ compilers poses intricate challenges,
+particularly when endeavoring to support an extensive set of CPU features across diverse
+architectures and multiple compilers. Furthermore, the conundrum of accommodating
+both fundamental features and supplementary features dispatched at runtime
+further complicates matters.
+
+In addition, the task of streamlining generic interface implementations often leads to the necessity
+of intricate solutions, which can obscure code readability and hinder debugging and maintenance.
+Nested namespaces, recursive sources, complex precompiled macros, or intricate meta templates
+are commonly employed but can result in convoluted code structures.
+
+Enter the proposed module, which offers a simple pragmatic multi-target solution to
+facilitate the management of CPU features with heightened ease and reliability.
```
This section reads more like an MR proposal, it feels like it's trying to justify the existence of the module. I don't know if that makes sense, and I'm not really sure how to improve it

dcbaker commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+# Copyright (c) 2023, NumPy Developers.
+# All rights reserved.
+import re
+from typing import (
+    Dict, Set, Tuple, List, Callable, Optional,
+    Union, Any, Iterable, cast, TYPE_CHECKING
+)
```
We don't do this anywhere else in Meson, we do `import typing as T` and then use the `T` namespace. For consistency I'd prefer if we kept with that.

dcbaker commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+        )
+        super().__init__(
+            opt_name, types,
+            convertor = lambda values: self.convert(
+                func_name, opt_name, values
+            ),
```
Using a lambda here feels wierd, could we just store the attributes against in the calss and then set `convertor = self.convert` as a bound class method?

dcbaker commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+
+        @typed_pos_args('features.new', str, int)
+        @typed_kwargs('features.new',
+            KwargInfo(
+                'implies',
+                (FeatureObject, ContainerTypeInfo(list, FeatureObject)),
+                default=[], listify=True
+            ),
+            KwargInfo(
+                'group', (str, ContainerTypeInfo(list, str)),
+                default=[], listify=True
+            ),
+            KwargConfilctAttr('features.new', 'detect', default=[]),
+            KwargConfilctAttr('features.new', 'args', default=[]),
+            KwargInfo('test_code', (str, File), default=''),
+            KwargInfo(
+                'extra_tests', (ContainerTypeInfo(dict, (str, File))),
+                default={}
+            ),
+            KwargInfo('disable', (str), default=''),
+        )
+        def init_attrs(state: 'ModuleState',
+                       args: Tuple[str, int],
+                       kwargs: 'FeatureKwArgs') -> None:
+            self.name = args[0]
+            self.interest = args[1]
+            self.implies = set(kwargs['implies'])
+            self.group = kwargs['group']
+            self.detect = kwargs['detect']
+            self.args = kwargs['args']
+            self.test_code = kwargs['test_code']
+            self.extra_tests = kwargs['extra_tests']
+            self.disable: str = kwargs['disable']
+            if not self.detect:
+                if self.group:
+                    self.detect = [ConflictAttr(val=f) for f in self.group]
+                else:
+                    self.detect = [ConflictAttr(val=self.name)]
+
+        init_attrs(state, args, kwargs)
```
Why is this being done here instead of in the caller?

dcbaker commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+# Copyright (c) 2023, NumPy Developers.
+# All rights reserved.
```
We've moved to using SPDX identifiers for the license, so we should have that there.

Also, this `All rights reserved` seems spurious, since to land in Meson it will need to be under the Apache-2.0 or a compatible permissive license like the MIT or BSD-2-clause

dcbaker commented on mesonbuild/modules/features/__init__.py:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2023, NumPy Developers.
+
```
I'd like to see `future.__annotations__` used as well, so we don't have to quote all of the argument types.

dcbaker reviewed (changes requested):
I haven't done a full review, but I have skimmed this and have some feedback.

Does numpy feel like this is baked enough to merge?

eli-schwartz commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+
+    """
+    val: str = field(hash=True, compare=True)
+    match: Union[re.Pattern, None] = field(
+        default=None, hash=False, compare=False
+    )
+    mfilter: Union[re.Pattern, None] = field(
+        default=None, hash=False, compare=False
+    )
+    mjoin: str = field(default='', hash=False, compare=False)
```
If we fit each assignment on one line, w won't exceed 88 chars, so we should do that.

eli-schwartz commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+            ret[attr] = val
+        return ret
+
+class KwargConfilctAttr(KwargInfo):
```
```suggestion
class KwargConflictAttr(KwargInfo):
```

(with suitable renames later in the file)

eli-schwartz commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+                continue
+            unknown_keys = [k for k in edict.keys() if k not in accepted_keys]
+            if unknown_keys:
+                raise MesonException(
+                    f'{func_name}: unknown keys {unknown_keys} in '
+                    f'option {opt_name}'
+                )
```
My feeling is that for ~100 characters it is still acceptable to be on one line when dealing with effectively just very long strings. Failing that, I'd still prefer to see:

```python
msg = 'long message here'
raise MesonException(msg)
```

eli-schwartz commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+            ret.append(implattr)
+        return ret
+
+if TYPE_CHECKING:
+    IMPLIED_ATTR = Union[
+        None, str, Dict[str, str], List[
+            Union[str, Dict[str, str]]
+        ]
+    ]
+    class FeatureKwArgs(TypedDict):
+        #implies: Optional[List['FeatureObject']]
+        implies: NotRequired[List[Any]]
+        group: NotRequired[List[str]]
+        detect: NotRequired[List[ConflictAttr]]
+        args: NotRequired[List[ConflictAttr]]
```
I would stick the typing logic up at the top, still.

ConflictAttr could be quoted in an `if T.TYPE_CHECKING:` block to cause it to parse as a forward reference, it doesn't *need* to be defined before.

eli-schwartz commented on mesonbuild/modules/features/feature.py:
```diff
@@ -0,0 +1,356 @@
+            KwargConfilctAttr('features.FeatureObject.update', 'args'),
+            KwargInfo('test_code', (NoneType, str, File)),
+            KwargInfo(
+                'extra_tests', (
+                    NoneType, ContainerTypeInfo(dict, (str, File)))
```
```suggestion
                'extra_tests', 
                (NoneType, ContainerTypeInfo(dict, (str, File)))),
```

These are two unrelated type arguments to KwargInfo.

eli-schwartz commented on mesonbuild/modules/features/utils.py:
```diff
@@ -0,0 +1,39 @@
+            )
+    return compiler
+
+def test_code(state: 'ModuleState', compiler: 'Compiler',
+              args: List[str], code: 'Union[str, File]'
+              ) -> Tuple[bool, bool, str]:
+    # TODO: Add option to treat warnings as errors
+    with compiler.cached_compile(
+        code, state.environment.coredata, extra_args=args
+    ) as p:
```
Same way that `) -> Tuple[bool, bool, str]:` is aligned with the function open parens, and not dedented to the beginning of the file, we should also have `with xxx():` either on the same line, or have it break across lines with an 8-space indent and no dedenting.

eli-schwartz commented on docs/markdown/Features-module.md:
```diff
@@ -0,0 +1,320 @@
+  str,
+  (
+    str | File | CustomTarget | CustomTargetIndex |
+    GeneratedList | StructuredSources | ExtractedObjects |
```
Are StructuredSources necessary here?

They are only permitted in the first place, in:
- Java jars
- rust targets (cannot be polyglot languages in the same target, you have to have a pure rust target that can be linked with other languages as a static/shared library)

So supporting it here feels premature.

eli-schwartz commented on mesonbuild/modules/features/module.py:
```diff
@@ -0,0 +1,711 @@
+
+    def gen_target(self, state: 'ModuleState', config_name: str,
+                   sources: List[Union[
+                      str, File, build.CustomTarget, build.CustomTargetIndex,
+                      build.GeneratedList, build.StructuredSources, build.ExtractedObjects,
+                      build.BuildTarget
```
```python
from ...interpreter.type_checking import SourcesVarargsType
```

gives you what you want here

eli-schwartz commented on mesonbuild/modules/features/module.py:
```diff
@@ -0,0 +1,711 @@
+    @permittedKwargs(build.known_stlib_kwargs | {
+        'dispatch', 'baseline', 'prefix', 'cached', 'keep_sort'
+    })
+    @typed_pos_args('features.multi_targets', str, min_varargs=1, varargs=(
+        str, File, build.CustomTarget, build.CustomTargetIndex,
+        build.GeneratedList, build.StructuredSources, build.ExtractedObjects,
+        build.BuildTarget
+    ))
```
Similarly here:

```python
from ...interpreter.type_checking import SOURCES_VARARGS
```

eli-schwartz commented on mesonbuild/modules/features/module.py:
```diff
@@ -0,0 +1,711 @@
+            target_name = test_result['target_name']
+            if target_name in enabled_targets_names:
+                skipped_targets.append((
+                    d, f'Dublicate target name "{target_name}"'
```
```suggestion
                    d, f'Duplicate target name "{target_name}"'
```

seiko2plus commented:
Hey everyone! Let me share what's going on with this module.

As you know, this was mainly built for NumPy during our switch to Meson. Back in early 2023, Jan Wassenberg (@jan-wassenberg) from Google and Christopher Sidebottom (@Mousius) from ARM suggested we use Highway instead of continuing with NumPy's own universal intrinsics, especially since NumPy had already started supporting C++.

Highway isn't just generic SIMD intrinsics - it also comes with a CPU Dispatcher that leverages C++ namespaces, metaprogramming, preprocessing, and CPU target attributes. It works on both modern Clang and GCC without any build system dependencies.

The approaches reflect different priorities:

- Our NumPy solution was all about compatibility
- Highway focuses more on portability

Right now, NumPy is moving to Highway as a library interface, and we've put NumPy's universal intrinsics on the shelf. We're temporarily using Meson CPU Dispatcher until we figure out whether to:

- Stick with Meson CPU Dispatcher for good
- Go all-in with Highway

If we end up sticking with Meson CPU Dispatcher, I think we should move all that Meson code from [numpy/meson_cpu](https://github.com/numpy/numpy/tree/main/meson_cpu) into this module (runtime library included) so other projects can benefit too. Looking at this code again after almost two years, it definitely needs some usability improvements. When I originally wrote it, I was just focused on getting SIMD optimization into NumPy as quickly as possible.

To be honest, I don't want to invest more time in this module until we make a final decision, so it probably won't be merged anytime soon. For now, I'm just focusing on fixing bugs and keeping things running through NumPy's vendored Meson version.

Sorry for the slow response to the pull request. I've had some tough personal stuff going on.

dcbaker commented:
@seiko2plus Given that status, would you be willing to set this module to a draft until you've decided what you want to do long term?

seiko2plus commented:
@dcbaker, Sure, no problems. Thank you.

seiko2plus committed:
Avoid using symlink, merge all the module tests into a single project

seiko2plus committed:
fix linting problems & make the module compatible with latest version
